### PR TITLE
Configuration

### PR DESCRIPTION
**Pull request by jules2689**
Constants -> Config

cc @richardwu

### CODE REVIEW

File: lib/news_scraper/extractors/google_news_rss.rb [hunk 1]
```diff
@@ -7,7 +7,8 @@ class GoogleNewsRss
       include ExtractorsHelpers
 
       BASE_URL = 'https://news.google.com/news?cf=all&hl=en&pz=1&ned=us&output=rss'.freeze
-      ARTICLE_URLS_DIR = Constants::TEMP_DIRS['extractors']['google_news_rss']['article_urls'].freeze
+      TEMP_DIRS = YAML.load_file(File.expand_path('../../../../config/temp_dirs.yml', __FILE__))
```
Comment (jules2689) [inline lib/news_scraper/extractors/google_news_rss.rb]: Temporary btw

File: lib/news_scraper/configuration.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module NewsScraper
+  class Configuration
+    DEFAULT_SCRAPE_PATTERNS_FILEPATH = File.expand_path('../../../config/article_scrape_patterns.yml', __FILE__)
+    attr_accessor :scrape_patterns, :scrape_patterns_filepath
+
+    def initialize(scrape_patterns_filepath: nil)
+      @scrape_patterns_filepath = scrape_patterns_filepath || DEFAULT_SCRAPE_PATTERNS_FILEPATH
+      raise ScrapePatternsFilePathDoesntExist unless File.exist?(@scrape_patterns_filepath)
```
Comment (richardwu) [inline lib/news_scraper/configuration.rb]: DoesNot

**Comment by richardwu** [comment]
Can we try using this design pattern? 

http://brandonhilkert.com/blog/ruby-gem-configuration-patterns/

**Comment by jules2689** [comment]
@richardwu re-ping

File: lib/news_scraper/configuration.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module NewsScraper
+  class Configuration
+    DEFAULT_SCRAPE_PATTERNS_FILEPATH = File.expand_path('../../../config/article_scrape_patterns.yml', __FILE__)
+    attr_accessor :scrape_patterns, :scrape_patterns_filepath
+
+    def initialize(scrape_patterns_filepath: nil)
+      self.scrape_patterns_filepath = scrape_patterns_filepath if scrape_patterns_filepath
+    end
+
+    def scrape_patterns_filepath=(file_path)
```
Comment (richardwu) [inline lib/news_scraper/configuration.rb:11]: Stick all of this logic in the `initialize` method and remove the params from https://github.com/richardwu/news_scraper/pull/30/files#diff-739722dce01e765226d9e3872d8388fbR6 ?

Comment (jules2689) [reply lib/news_scraper/configuration.rb:11]: We can't delete this. This is what makes the assignment in the block work

Comment (richardwu) [reply lib/news_scraper/configuration.rb:11]: True.

File: lib/news_scraper/configuration.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
+module NewsScraper
+  class Configuration
+    DEFAULT_SCRAPE_PATTERNS_FILEPATH = File.expand_path('../../../config/article_scrape_patterns.yml', __FILE__)
+    attr_accessor :scrape_patterns, :scrape_patterns_filepath
```
Comment (richardwu) [inline lib/news_scraper/configuration.rb]: I don't like how `scrape_patterns` can also be set without regard to the `scrape_patterns_filepath`.

You could end up with a `scrape_patterns` that isn't the same as the scrape patterns in `scrape_patterns_filepath`.

Only one of these should be writable too, and that's probably `scrape_patterns_filepath`.

File: lib/news_scraper/configuration.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    def initialize(scrape_patterns_filepath: nil)
+      self.scrape_patterns_filepath = scrape_patterns_filepath if scrape_patterns_filepath
```
Comment (richardwu) [inline lib/news_scraper/configuration.rb]: You need to replace this with 

```
@scrape_patterns_filepath = DEFAULT_SCRAPE_PATTERNS_FILEPATH
```

Or else this won't be set at all during initialization.

File: lib/news_scraper/configuration.rb (hunk 4 above)
Comment (jules2689) [reply lib/news_scraper/configuration.rb]: true.

Comment (jules2689) [reply lib/news_scraper/configuration.rb]: Well, in a future iteration - it might be one or the other is settable since scraping doesnt need a filepath

File: lib/news_scraper/configuration.rb (hunk 5 above)
Comment (jules2689) [reply lib/news_scraper/configuration.rb]: that's not gonna fly either because of

```
  def configuration
    @configuration ||= Configuration.new
  end

  def reset_configuration
    @configuration = Configuration.new
  end
```

Comment (jules2689) [reply lib/news_scraper/configuration.rb]: actually, I suppose it would just be overwritten, nvm

Comment (richardwu) [reply lib/news_scraper/configuration.rb]: It's a new object so yea.

File: lib/news_scraper/configuration.rb [hunk 6]
```diff
@@ -0,0 +1,17 @@
+module NewsScraper
+  class Configuration
+    DEFAULT_SCRAPE_PATTERNS_FILEPATH = File.expand_path('../../../config/article_scrape_patterns.yml', __FILE__)
+    attr_accessor :scrape_patterns_filepath
+    attr_reader :scrape_patterns
+
+    def initialize(scrape_patterns_filepath: DEFAULT_SCRAPE_PATTERNS_FILEPATH)
```
Comment (richardwu) [inline lib/news_scraper/configuration.rb]: Can we keep this initialize method with empty params? YAGNI

File: test/test_helper.rb [hunk 7]
```diff
@@ -9,38 +9,24 @@
 
 module MiniTest
   class Test
+    def setup
+      super
+
+      path = "/tmp/#{location.tr('#', '_')}"
```
Comment (richardwu) [inline test/test_helper.rb]: What is `location`?

File: test/test_helper.rb [hunk 8]
```diff
@@ -9,38 +9,24 @@
...
+      FileUtils.touch(path)
+      File.write(path, File.read(NewsScraper::Configuration::DEFAULT_SCRAPE_PATTERNS_FILEPATH))
```
Comment (richardwu) [inline test/test_helper.rb]: Why don't you just use `NewsScraper.configuration.scrape_patterns`?

File: test/unit/config/article_scrape_patterns_test.rb [hunk 9]
```diff
@@ -4,7 +4,7 @@ class ArticleScrapePatternsTest < Minitest::Test
   VALID_METHODS = %w(css xpath readability).freeze
 
   def setup
-    @scrape_patterns = YAML.load_file(NewsScraper::Constants::SCRAPE_PATTERN_FILEPATH)
+    @scrape_patterns = YAML.load_file(NewsScraper::Configuration::DEFAULT_SCRAPE_PATTERNS_FILEPATH)
```
Comment (richardwu) [inline test/unit/config/article_scrape_patterns_test.rb]: `NewsScraper.configuration.scrape_patterns`?

File: test/unit/config/article_scrape_patterns_test.rb [hunk 10]
```diff
@@ -57,7 +57,9 @@ def test_scrape_methods_presets_are_valid
   end
 
   def test_scrape_methods_have_proper_variable_names
-    article_scrape_patterns = File.read(NewsScraper::Constants::SCRAPE_PATTERN_FILEPATH).split("\n").map(&:strip)
+    article_scrape_patterns = File.read(NewsScraper::Configuration::DEFAULT_SCRAPE_PATTERNS_FILEPATH)
```
Comment (richardwu) [inline test/unit/config/article_scrape_patterns_test.rb:61]: `@scrape_patterns`

File: test/unit/config/xpath_css_path_test.rb [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 class XpathCssPathTest < Minitest::Test
   def setup
-    @scrape_patterns = YAML.load_file('config/article_scrape_patterns.yml')
+    @scrape_patterns = YAML.load_file(NewsScraper::Configuration::DEFAULT_SCRAPE_PATTERNS_FILEPATH)
```
Comment (richardwu) [inline test/unit/config/xpath_css_path_test.rb]: `NewScraper.configuration.scrape_patterns`

File: test/unit/configuration_test.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+        Configuration.new(scrape_patterns_filepath: "nope")
+      end
+    end
+
+    def test_scrape_patterns_loaded_from_filepath
+      tmp_file = Tempfile.new('test_scrape_patterns_loaded_from_filepath')
+      tmp_file.write("domains:\n  test")
+      tmp_file.rewind
+
+      assert_equal({ 'domains' => 'test' }, Configuration.new(scrape_patterns_filepath: tmp_file.path).scrape_patterns)
```
Comment (richardwu) [inline test/unit/configuration_test.rb]: I don't think you should test passing in through params (see above comment) since this isn't how it works in the actual code.

File: test/unit/configuration_test.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    end
+
+    def test_configuration_block
+      NewsScraper.configure do |config|
+        config.scrape_patterns_filepath = Tempfile.new('Tempfile')
+      end
```
Comment (richardwu) [inline test/unit/configuration_test.rb]: Move all your logic from the test above to here, but instead of passing it in through the params using the configure block.

**Review by richardwu — CHANGES_REQUESTED** [review]

File: test/test_helper.rb (hunk 7 above)
Comment (jules2689) [reply test/test_helper.rb]: location is a MiniTest method to get the currently tested method. It's just a way to get a unique tmp path. `Tempfile` seemed to be garbage collected too soon :(

File: test/test_helper.rb (hunk 8 above)
Comment (jules2689) [reply test/test_helper.rb]: wasnt like that before ;)

File: test/unit/config/article_scrape_patterns_test.rb (hunk 9 above)
Comment (jules2689) [reply test/unit/config/article_scrape_patterns_test.rb]: This would load the tempfile one, which should be okay. I had wanted to be explicit about loading the default one though.

File: test/unit/configuration_test.rb (hunk 13 above)
Comment (jules2689) [reply test/unit/configuration_test.rb]: ugh, I forgot to finish this test. Was tired D;

File: test/unit/config/article_scrape_patterns_test.rb (hunk 10 above)
Comment (jules2689) [reply test/unit/config/article_scrape_patterns_test.rb:61]: Can't do that. This test is asserting that the assumed variable names are correct. e.g. `&class_author`

File: test/unit/config/article_scrape_patterns_test.rb (hunk 9 above)
Comment (jules2689) [reply test/unit/config/article_scrape_patterns_test.rb]: This is only ok if we do a `reset_configuration` in the global minitest setup. Because this is a global class variable now, it leaks between tests.

**Comment by jules2689** [comment]
Thoughts on extending `Forwardable` in configuration, and then delegating `[]` access to `scrape_patterns` with `def_delegator :scrape_patterns, :[]`

That means `NewsScraper.configuration.scrape_patterns['data_types']` would now go to `NewsScraper.configuration['data_types']`

**Comment by jules2689** [comment]
Reping @richardwu

File: lib/news_scraper/configuration.rb [hunk 14]
```diff
@@ -0,0 +1,17 @@
+module NewsScraper
+  class Configuration
+    DEFAULT_SCRAPE_PATTERNS_FILEPATH = File.expand_path('../../../config/article_scrape_patterns.yml', __FILE__)
+    attr_accessor :scrape_patterns_filepath
+    attr_reader :scrape_patterns
+
+    def initialize
+      self.scrape_patterns_filepath = DEFAULT_SCRAPE_PATTERNS_FILEPATH
```
Comment (richardwu) [inline lib/news_scraper/configuration.rb]: Why are you using `self` instead of `@`?

File: test/test_helper.rb [hunk 15]
```diff
@@ -9,38 +9,26 @@
...
-    def raw_data_fixture(domain)
-      File.read("test/data/articles/#{domain.tr('.', '_')}_raw")
-    end
+    def setup
+      super
 
-    def transformation_fixture(domain)
-      YAML.load_file("test/data/articles/#{domain.tr('.', '_')}_transformed.yml")
-    end
+      NewsScraper.reset_configuration
```
Comment (richardwu) [inline test/test_helper.rb:15]: Why is this necessary?

File: test/test_helper.rb [hunk 16]
```diff
@@ -9,38 +9,26 @@
...
 
-    def stub_temp_file_with_path(file_path)
-      original_content = File.read(file_path)
-      # Use a tmp dir so as not to override the actual config/article_scrape_patterns.yml
-      Dir.mktmpdir do |dir|
-        temp_path = File.join(dir, file_path)
-        FileUtils.mkpath(File.dirname(temp_path))
+      path = "/tmp/#{location.tr('#', '_')}"
+      FileUtils.touch(path)
+      File.write(path, NewsScraper.configuration.scrape_patterns.to_yaml)
```
Comment (richardwu) [inline test/test_helper.rb]: Wouldn't this file persist after the tests?

File: test/unit/transformers/article_test.rb [hunk 17]
```diff
@@ -1,30 +1,34 @@
 require 'test_helper'
 
-class ArticleTest < Minitest::Test
-  def test_transform_returns_correct_json_transformation_for_supported_domains
-    supported_domains = YAML.load_file('config/article_scrape_patterns.yml')['domains'].keys
+module NewsScraper
+  module Transformers
+    class ArticleTest < Minitest::Test
+      def test_transform_returns_correct_json_transformation_for_supported_domains
+        supported_domains = YAML.load_file(Configuration::DEFAULT_SCRAPE_PATTERNS_FILEPATH)['domains'].keys
```
Comment (richardwu) [inline test/unit/transformers/article_test.rb]: `configuration.scrape_patterns`

**Review by richardwu — CHANGES_REQUESTED** [review]

**Comment by richardwu** [comment]
> Thoughts on extending Forwardable in configuration, and then delegating [] access to scrape_patterns with def_delegator :scrape_patterns, :[]

That means NewsScraper.configuration.scrape_patterns['data_types'] would now go to NewsScraper.configuration['data_types']

Not a big fan since it's ambiguous what `data_types` it's referring to in the gem. Esp. since `articles` could only be a subset of the things we support.

File: test/test_helper.rb (hunk 15 above)
Comment (jules2689) [reply test/test_helper.rb:15]: Cause NewsScraper configuration is a class global whatever variable and not resetting will make this leak into new tests and then "kaboom"

File: test/test_helper.rb (hunk 16 above)
Comment (jules2689) [reply test/test_helper.rb]: technically yes it will persist, but it's in the system /tmp dir, I can remove

Comment (jules2689) [reply test/test_helper.rb]: fixed.

File: test/unit/transformers/article_test.rb (hunk 17 above)
Comment (jules2689) [reply test/unit/transformers/article_test.rb]: done
